Tests/Client/Unit Tests/Client+DevicesTests.swift

```diff
@@ -0,0 +1,109 @@
+//
+//  Client+DevicesTests.swift
+//  StreamChatClientTests
+//
+//  Copyright © 2020 Stream.io Inc. All rights reserved.
+//
+
+import XCTest
+
+@testable import StreamChatClient
+
+class Client_DevicesTests: XCTestCase {
+
+    // Ideally, we would create a new client for every test case. Unfortunately, this is not technically
+    // possible at the moment.
+    lazy var client: Client = {
+        let sessionConfig = URLSessionConfiguration.default
+
+        sessionConfig.protocolClasses?.insert(RequestRecorderURLProtocol.self, at: 0)
+        Client.config = .init(apiKey: "test_api_key", defaultURLSessionConfiguration: sessionConfig)
+        return Client.shared
+    }()
+
+    func test_getDevice_createsRequest() {
+        // Setup
+        let testUser = User(id: "test_user_\(UUID())")
+        client.set(user: testUser, token: "test_token")
+
+        // Action
+        client.devices { _ in }
+
+        // Assert
+        AssertNetworkRequest(
+            method: .get,
+            path: "/devices",
+            headers: ["Content-Type": "application/json"],
+            queryParameters: ["api_key": "test_api_key"],
+            body: nil
+        )
```

Comment:
How about we embed the statement making the network call in this function?
```swift
func AssertNetworkRequest(
    action: @autoclosure () -> Any,
    method: Endpoint.Method,
    path: String,
    headers: [String: String]?,
    queryParameters: [String: String]?,
    body: [String: Any]?,
    timeout: TimeInterval = 0.5,
    file: StaticString = #file,
    line: UInt = #line
)
```
I think it's more inline with XCTest functions with this approach. We can use it with:
```swift
func test_getDevice_createsRequest() {
    // Setup
    let testUser = User(id: "test_user_\(UUID())")
    client.set(user: testUser, token: "test_token")

    // Assert
    AssertNetworkRequest(
        action: client.devices { _ in },
        method: .get,
        path: "/devices",
        headers: ["Content-Type": "application/json"],
        queryParameters: ["api_key": "test_api_key"],
        body: nil
    )
}
```
by adding `_ = action()` as the first line in `func AssertNetworkRequest`. The action and the expected result are coupled with this way.

Reply:
I understand what you're trying to achieve there. I see two reasons for not doing that:

- The request is not a result of the action but its side effect. Using it directly in the assertion implies the `action` signature is something like `() -> Request`.

- When the action is more than a couple of lines, it becomes hard to parse it visually and you'd probably refactor it into a variable anyway:
```swift
func test_getDevice_createsRequest() {
    // Setup
    let testUser = User(id: "test_user_\(UUID())")
    client.set(user: testUser, token: "test_token")

    let action = { 
       .... 
    }

    // Assert
    AssertNetworkRequest(
        action: action,
        method: .get,
        path: "/devices",
        headers: ["Content-Type": "application/json"],
        queryParameters: ["api_key": "test_api_key"],
        body: nil
    )
}
```

Reply:
This is just the first version of this function. I'm sure we'll have a lot of improvements and ideas once we start using it extensively on day-to-day basis.

Reply:
I don't agree with the points:
* Network request is a direct result of the call being made, just not the return value.
* A network request will be a result of a single call (of client, network, etc..) so if it requires more than a line, it can be boiled down a single call.

But I definitely agree that this is the first implementation and we'll improve, we can keep it like this if you think so.

Reply:
`Cancellable` is the return value of this, everything else is technically a side effect.

If you don't mind I'd keep it as is. My arguments are not strong enough to convince you, nor are your to convince me. Time will tell ... :)

Reply:
Thank you for explaining, since I don't have any more arguments let's keep it as it is 👍🏼

Reply:
Both suggestions look fine for me.